### ISSUE TITLE
Look into adding a CMM primitive for non-reordable read

### ISSUE THREAD

**Issue by rrnewton** [iu-parfunc/sc-haskell#5]
We need more research into CMM to figure out where this would go.  @osa1 do you have experience with that?

**Comment by osa1** [comment]
Yes, I've added a primop to GHC before. What exactly do you need?

**Comment by rrnewton** [comment]
Well, maybe its just a primop... and maybe it's a modification to the CMM AST.  I need to learn the CMM AST.

Our current understanding is that `readIORef`/`readMutVar#` becomes a normal memory read at CMM level (and hence can get reordered).  Is this true?

If so maybe the solution is to leave it as a primop until codegen...  as implied in the name of this repo the goal is to test the overhead of sequential consistency for Haskell, which means locking down reordering of readIORef.

(One we do need to compiler readSTRef differently.. so this may mean introducing a different primop for that distinction.  Not sure yet; different ways to go.)

**Comment by madanMus** [comment]
What is CMM?

**Comment by rrnewton** [comment]
"C minus minus".  It's the [third of three major IRs in the GHC compiler](https://ghc.haskell.org/trac/ghc/wiki/Commentary/Compiler): "Core", "STG", "CMM".

We were reminded [when rereading an old blog post](http://blog.ezyang.com/2014/01/so-you-want-to-add-a-new-concurrency-primitive-to-ghc/) that there may be a CMM optimization pass that's moving reads, which would then have to contend with to make sure it doesn't move our IORef reads.
